src/java/org/apache/cassandra/db/filter/IndexHints.java

```diff
@@ -0,0 +1,437 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.db.filter;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import com.google.common.collect.Sets;
+
+import org.apache.cassandra.cql3.QualifiedName;
+import org.apache.cassandra.db.TypeSizes;
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.index.Index;
+import org.apache.cassandra.index.IndexRegistry;
+import org.apache.cassandra.io.util.DataInputPlus;
+import org.apache.cassandra.io.util.DataOutputPlus;
+import org.apache.cassandra.locator.InetAddressAndPort;
+import org.apache.cassandra.net.MessagingService;
+import org.apache.cassandra.schema.IndexMetadata;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.utils.FBUtilities;
+
+import static java.lang.String.format;
+
+/**
+ * User-provided directives about what indexes should be used by a {@code SELECT} query. It consists of a set of indexes
+ * that should be preferred and a set of indexes that should not be used. Other than that, indexes that are applicable
+ * to the query and that are not mentioned on these two sets might or might not be used depending on the index query
+ * planner.
+ */
+public class IndexHints
+{
+    static final String CONFLICTING_INDEXES_ERROR = "Indexes cannot be both preferred and excluded: ";
+    static final String WRONG_KEYSPACE_ERROR = "Index %s is not in the same keyspace as the queried table.";
+    static final String MISSING_INDEX_ERROR = "Table %s doesn't have an index named %s";
+
+    public static final IndexHints NONE = new IndexHints(Collections.emptySet(), Collections.emptySet());
+
+    public static final Serializer serializer = new Serializer();
+
+    /**
+     * The indexes to prefer when executing a query.
+     */
+    public final Set<IndexMetadata> preferred;
+
+    /**
+     * The indexes not to use when executing the query.
+     */
+    public final Set<IndexMetadata> excluded;
```

Comment:
Suggestion: `avoided`? 
Excluded suggests it is very strong, strict - so the indexes from this list would never be used.
The meaning of "avoided" seems to balance more with "preferred" (which is also not strong).

Reply:
Now after reading the rest of the implementation, I can see that they *are* indeed excluded. So this is fine to stay like that, but maybe we should write in the docs that `excluded` means always excluded and not just a hint?